### PR TITLE
[classes] Some refactoring on proof save preparation.

### PR DESCRIPTION
**Pull request by ejgallego**
Note the ugly problem that we have on close_proof:
```ocaml
proof_global.ml:278

  let entry_fn p (_, t) =
    let t = EConstr.Unsafe.to_constr t in
    let univstyp, body = make_body t p in
```
arguments of start_proof should be pre-normalized.

I think this will require a lot of refactoring to be fixed properly.

### CODE REVIEW

File: vernac/classes.ml [hunk 1]
```diff
@@ -516,7 +513,7 @@ let do_instance_program env env' sigma ?hook ~global ~poly cty k u ctx ctx' pri
       Some term, termtype, sigma in
   let termtype, sigma = do_instance_resolve_TC term termtype sigma env in
   if not (Evd.has_undefined sigma) && not (Option.is_empty opt_props) then
-    let term = to_constr sigma (Option.get term) in
+    let term = Option.get term in
```
Comment (SkySkimmer) [inline vernac/classes.ml]: While you're at it you could remove the option wrapping here

File: vernac/classes.ml [hunk 2]
```diff
@@ -353,11 +348,12 @@ let declare_instance_program env sigma ~global ~poly id pri imps decl term termt
   let obls, constr, typ =
     match term with
     | Some t ->
-      let termtype = EConstr.of_constr termtype in
       let obls, _, constr, typ =
         Obligations.eterm_obligations env id sigma 0 t termtype
       in obls, Some constr, typ
-    | None -> [||], None, termtype
+    | None ->
+      let termtype = EConstr.to_constr sigma termtype in
```
Comment (SkySkimmer) [inline vernac/classes.ml]: Why are we normalizing here now?

Comment (ejgallego) [reply vernac/classes.ml]: We now pass an `EConstr.t` to the function; we are more lazy this way I think.

File: vernac/classes.ml (hunk 1 above)
Comment (ejgallego) [reply vernac/classes.ml]: Indeed, that was a good idea, done.

File: vernac/classes.ml [hunk 3]
```diff
@@ -345,20 +345,13 @@ let declare_instance_program env sigma ~global ~poly id pri imps decl term termt
...
-    | None ->
-      let termtype = EConstr.to_constr sigma termtype in
-      [||], None, termtype
-  in
+  let obls, _, term, typ = Obligations.eterm_obligations env name sigma 0 term termtype in
   let hook = DeclareDef.Hook.make hook in
   let ctx = Evd.evar_universe_context sigma in
-  ignore(Obligations.add_definition ~name:id ?term:constr
-           ~univdecl:decl ~scope:(DeclareDef.Global Declare.ImportDefaultBehavior) ~poly ~kind:Decls.Instance ~hook typ ctx obls)
+  let scope, kind = DeclareDef.Global Declare.ImportDefaultBehavior, Decls.Instance in
```
Comment (SkySkimmer) [inline vernac/classes.ml]: Why tuple?

File: vernac/classes.ml [hunk 4]
```diff
@@ -314,15 +314,10 @@ let instance_hook info global imps ?hook cst =
...
-                                 (CVars.universes_of_constr term) in
-    Evd.restrict_universe_context sigma levels
-  in
-  let uctx = Evd.check_univ_decl ~poly sigma decl in
   let kind = Decls.(IsDefinition Instance) in
-  let entry = Declare.definition_entry ~types:termtype ~univs:uctx term in
+  (* XXX: Type needs to be free of evars, not the body tho *)
+  let sigma, entry = DeclareDef.prepare_definition
+      ~allow_evars:true ~poly sigma decl ~types:(Some termtype) ~body:term in
   let kn = Declare.declare_constant ~name ~kind (Declare.DefinitionEntry entry) in
```
Comment (SkySkimmer) [inline vernac/classes.ml]: Won't declare_constant die if the body has evars?

**Comment by SkySkimmer** [comment]
>which means than arguments of start_proof should be pre-normalized.

Why can't we call nf_evar in proofview.init?

File: vernac/classes.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/classes.ml]: I was just lazy and saved a line.

File: vernac/classes.ml (hunk 4 above)
Comment (ejgallego) [reply vernac/classes.ml]: Good question, when I wrote this I definitively needed the `allow_evars:true`, let me recall the details.

**Comment by ejgallego** [comment]
> Why can't we call nf_evar in proofview.init?

We can, but I'm afraid of performance problems; the right fix seems to carry an `EConstr`; even if we do so, we still have to normalize again at the end as the term may have gotten instantiated [in fact I recall that was the problem]

File: vernac/classes.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/classes.ml]: I left it as is, if you think it is a big deal let me know and I'll split the line.

**Review by SkySkimmer — APPROVED** [review]
We end up doing univ minimization twice because of do_instance_resolve_TC, hopefully we can reduce such duplication eventually.

**Comment by ejgallego** [comment]
> We end up doing univ minimization twice because of do_instance_resolve_TC, hopefully we can reduce such duplication eventually.

How can we solve that? Should we add a parameter to `do_instance_resolve_TC`, or just move the minimization to `declare_instance_open`? Where is the minimization actually needed in the interactive case?

[I hope to centralize the non-interactive part by moving finalize to be part of `Declare` and abstracting the type of the entries]

**Pull request merged**